File: Audio Inpainting/utils/makeClippedSignal.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import spectrogram, hann
from scipy.io.wavfile import write, read
from typing import Tuple, Optional

def makeClippedSignal(
    x: np.ndarray, clippingLevel: float, 
    GR: Optional[bool] = False
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, Optional[np.ndarray]]:
    
    """
    Normalize and clip a signal.

    Args:
        x (np.ndarray): Input signal (may be multichannel).
        clippingLevel (float): Clipping level, between 0 and 1.
        GR (bool, optional): Flag to generate an optional graphical display. Defaults to False.

    Returns:
        Tuple[np.ndarray, np.ndarray, np.ndarray, Optional[np.ndarray]]: A tuple of numpy arrays.
            - xClipped (np.ndarray): Clipped signal.
            - IClipped (np.ndarray): Boolean vector (same size as xClipped) that indexes clipped samples.
            - xClean (np.ndarray): Clean signal.
            - clipSizes (np.ndarray, optional): Size of the clipped segments. Only returned if GR is True.
    """
    pass

    # Normalize and clip a signal.
    xMax = 0.9999
    xClean = x / np.max(np.abs(x)) * xMax
    clippingLevel = clippingLevel * xMax

    # Clipping (hard threshold)
    xClipped = np.minimum(np.maximum(xClean, -clippingLevel), clippingLevel)
    IClipped = np.abs(xClipped) >= clippingLevel  # related indices

    # Size of the clipped segments
    if GR:
        clipSizes = np.diff(IClipped.astype(int))
        if clipSizes[np.where(clipSizes)[0][0]] == -1:
            clipSizes = np.concatenate(([1], clipSizes))
        if clipSizes[np.where(clipSizes)[0][-1]] == 1:
            clipSizes = np.concatenate((clipSizes, [-1]))
        clipSizes = np.diff(np.where(clipSizes))
        clipSizes = clipSizes[::2]

    # Optional graphical display
    if GR:
        # Plot histogram of the sizes of the clipped segments
        plt.figure()
        plt.hist(clipSizes, bins=np.arange(1, np.max(clipSizes)+1))
        plt.title('Size of missing segments')
        plt.xlabel('Size')
        plt.ylabel('# of segments')

        t = np.arange(len(xClean))  # time scale in samples

        # Plot original and clipped signals
        plt.figure()
        plt.plot(t, xClean, label='original')
        plt.plot(t, xClipped, label='clipped')
        plt.legend()

        # Scatter plot between original and clipped signals
        plt.figure()
        plt.plot(xClean, xClipped, '.')
        plt.xlabel('Original signal')
        plt.ylabel('Clipped signal')

        # Spectrograms
        N = 512
        w = hann(N)
        fs = 1
        NOverlap = round(.8 * N)
        nfft = 2 ** (int(np.log2(N)) + 2)

        fig, axs = plt.subplots(3, 3)
        axs[0, 0].specgram(xClean, NFFT=N, Fs=fs, window=w, noverlap=NOverlap, mode='magnitude')
        axs[0, 0].set_title('Original')
        axs[0, 1].specgram(xClipped, NFFT=N, Fs=fs, window=w, noverlap=NOverlap, mode='magnitude')
        axs[0, 1].set_title('Clipped')
        axs[0, 2].specgram(xClean - xClipped, NFFT=N, Fs=fs, window=w, noverlap=NOverlap, mode='magnitude')
        axs[0, 2].set_title('Error (=original-clipped)')
        axs[1, 0].plot(t, xClean)
        axs[1, 1].plot(t, xClean, t, xClipped)
        axs[1, 2].plot(t, xClean - xClipped)

    return xClipped, IClipped, xClean

File: AudioInpaintingToolbox/Utils/makeClippedSignal.m
function [xClipped, IClipped, xClean, clipSizes] = makeClippedSignal(x,clippingLevel,GR)
% Normalize and clip a signal.
%
% Usage:
%   [xClipped, IClipped, xClean, clipSizes] = makeClippedSignal(x,clippingLevel,GR)
%
% Inputs:
%   - x: input signal (may be multichannel)
%   - clippingLevel: clipping level, between 0 and 1
%   - GR (default: false): flag to generate an optional graphical display
%
% Outputs:
%   - xClipped: clipped signal
%   - IClipped: boolean vector (same size as xClipped) that indexes clipped
%   samples
%   - xClean: clean signal
%   - clipSizes: size of the clipped segments
%
% Note that the input signal is normalized to 0.9999 (-1 is not allowed in
% wav files) to provide xClipped and xClean.
%
% -------------------
%
% Audio Inpainting toolbox
% Date: June 28, 2011
% By Valentin Emiya, Amir Adler, Maria Jafari
% This code is distributed under the terms of the GNU Public License version 3 (http://www.gnu.org/licenses/gpl.txt).

if nargin<3 || isempty(GR)
    GR = false;
end

%% Normalization
xMax = 0.9999;
xClean = x/max(abs(x(:)))*xMax;
clippingLevel = clippingLevel*xMax;

%% DISABLED - Ramp to produce a clipping level that linearly increases
if 0
    xClean = xClean.*(1:length(xClean))'/length(xClean);
end

%% Clipping (hard threshold)
xClipped = min(max(xClean,-clippingLevel),clippingLevel);
IClipped = abs(xClipped)>=clippingLevel; % related indices

%% Size of the clipped segments
if nargout>3 || GR
   %     clipSizes = diff(find(diff(~IClipped)));
   %     clipSizes = clipSizes(2-(IClipped(1)==0):2:end);
   clipSizes = diff(IClipped);
   if clipSizes(find(clipSizes,1,'first'))==-1,clipSizes = [1;clipSizes]; end
   if clipSizes(find(clipSizes,1,'last'))==1,clipSizes = [clipSizes;-1]; end
   clipSizes = diff(find(clipSizes));
   clipSizes = clipSizes(1:2:end);
end

%% Optional graphical display
if GR
    
    % Plot histogram of the sizes of the clipped segments
    if ~isempty(clipSizes)
        figure
        hist(clipSizes,1:max(clipSizes))
        title('Size of missing segments')
        xlabel('Size'),ylabel('# of segments')
    end
    
    t = (0:length(xClean)-1); % time scale in samples
    
    % Plot original and clipped signals
    figure
    plot(t,xClean,'',t,xClipped,'')
    legend('original','clipped')
    
    % Scatter plot between original and clipped signals
    figure
    plot(xClean,xClipped,'.')
    xlabel('Original signal'),ylabel('Clipped signal')
    
    % Spectrograms
    N = 512;
    w = hann(N);
    fs = 1;
    NOverlap = round(.8*N);
    nfft = 2^nextpow2(N)*2*2;
    figure
    subplot(3,3,[1,4])
    spectrogram(xClean,w,NOverlap,nfft,fs,'yaxis')
    title('Original')
    xlim(t([1,end]))
    cl = get(gca,'clim');
    set(gca,'clim',cl);
    subplot(3,3,[1,4]+1)
    spectrogram(xClipped,w,NOverlap,nfft,fs,'yaxis')
    title('Clipped')
    set(gca,'clim',cl);
    subplot(3,3,[1,4]+2)
    spectrogram(xClean-xClipped,w,NOverlap,nfft,fs,'yaxis')
    title('Error (=original-clipped)')
    set(gca,'clim',cl);
    subplot(3,3,7)
    plot(t,xClean,'');xlim(t([1,end]))
    subplot(3,3,8)
    plot(t,xClean,'',t,xClipped,'');xlim(t([1,end]))
    subplot(3,3,9)
    plot(t,xClean-xClipped,'');xlim(t([1,end]))
end

return
